### PR TITLE
support case list forms in child modules

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?178088

@TylerSheffels @dannyroberts 

review by commit

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml.py [hunk 1]
```diff
@@ -898,42 +904,71 @@ def get_case_type_created_by_form(form):
...
-                        except SuiteError:
-                            if source_meta.requires_selection:
-                                raise
+                            [target_dm] = [
+                                target_meta for target_meta in target_form_dm
+                                if target_meta.case_type == case_type
+                            ]
+                        except ValueError:
+                            raise SuiteError(
+                                "Return module for case list form has mismatching datums: {}".format(form.unique_id)
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 2]
```diff
@@ -228,46 +152,54 @@ def test_case_list_form_parent_child_basic(self):
...
+
+    def test_case_list_form_parent_child_submodule(self):
+        """
+        * Register house (case type = house, basic)
+          * Register house form
+        * Register person (case type = person, parent select = 'Register house', advanced)
+          * Register person form
+        * Update house (case type = house, case list form = 'Register house')
+          * Update house form
+          * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (129 > 115 characters)

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 3]
```diff
@@ -228,46 +152,54 @@ def test_case_list_form_parent_child_basic(self):
...
+
+        person_module, update_person_form = factory.new_advanced_module('update_person', 'person')
+        person_module.case_list_form.form_id = register_person_form.unique_id
+        person_module.root_module_id = house_module.unique_id
+
+        factory.form_updates_case(update_person_form, 'house')
+        factory.form_updates_case(update_person_form, 'person', parent_case_type='house')
 
-        self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-basic'), app.create_suite())
+        self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-submodule'), factory.app.create_suite())
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (116 > 115 characters)

**Comment by dannyroberts** [comment]
:+1:

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/tests/test_case_list_form.py, line 179 - E501 line too long (126 > 115 characters)
* corehq/apps/app_manager/tests/test_case_list_form.py, line 200 - E501 line too long (122 > 115 characters)
* corehq/apps/app_manager/tests/test_case_list_form.py, line 233 - E501 line too long (125 > 115 characters)
* corehq/apps/app_manager/tests/test_case_list_form.py, line 243 - E501 line too long (129 > 115 characters)
* corehq/apps/app_manager/tests/test_case_list_form.py, line 265 - E501 line too long (122 > 115 characters)

**Comment by snopoke** [comment]
@dannyroberts this should be GTG once tests pass. Latest change is adding more tests and a small tweak or two.

**Comment by dannyroberts** [comment]
@snopoke great, @TylerSheffels and I are going to go over this together in a couple hours, and then I'll merge

**Comment by snopoke** [comment]
👍

**Comment by dannyroberts** [comment]
@snopoke these are getting pretty tricky to follow, even when the changes are small. Would you be able to describe to me our overall approach to case list forms and the other things it touches? Did this PR add one more special case, or did it simplify the orignal code to follow a more general pattern?

**Comment by snopoke** [comment]
Absolutely, I was thinking on the way home that I need to do a write-up on
this stuff. I'll ping you when I've got something.

This PR is what all the others have been building up to. The full use case
is to be able to have a case list form in a submodule whose case type is a
child of the parent modules case type.

There is also more refactoring that I'd like to do which will hopefully
make things easier to follow. But I don't know when I'll get to that.
Hopefully before the end of the year!

**Pull request merged**